### PR TITLE
docs: tighten spec ↔ persona ↔ docs layering (audit-driven)

### PR DESCRIPTION
**Pull request by 0xbulma**
> **Stacked on #682.** Once #682 merges to `main`, this diff will collapse to only the changes below. The audit-driven cleanup spans 5 commits — see commit history at the bottom.

## Motivation

After #682 (which extracted personas + added `ci-release-security`), four audit rounds surfaced gaps in the documentation architecture:

1. **`ci-release-security` had no `AGENTS.md` anchor.** The persona enforced CI/release rules — workflow injection, action pinning, publish-flow integrity, lockfile drift — but those rules weren't codified anywhere in the spec. The persona was effectively source-of-truth, inverting the layering.
2. **The `documentation` persona only covered JSDoc on exports.** Markdown docs (`README`, `AGENTS.md`, `docs/**`, `.agents/**`) could drift from the code; cross-references could break on file moves; AGENTS.md ↔ persona backlinks could fall out of sync — none of it caught by review.
3. **No navigation from rules to enforcers.** A rule-author editing `AGENTS.md` §3 had no way to find which personas needed updating.
4. **Persona overlap / verbatim duplication.** Several personas restated AGENTS.md rules verbatim instead of pointing at the section — sync burden + drift risk. Overlapping concerns between `code-quality`/`silent-failure-hunter`, `module-api-architecture`/`style-conventions`, `code-quality`/`web3-security`.
5. **`test-coverage` persona was out of date** — described the obsolete `packages/<pkg>/test/` layout while AGENTS.md §5 mandates colocation (`src/Foo.ts` ↔ `src/Foo.test.ts`). Doesn't reflect the incoming TIB-2026-04-27 (PR #596) on transport-level mocking.
6. **CONTRIBUTING.md duplicated AGENTS.md** §7 (changesets) and §8 (code style) — sync burden.
7. **`docs/tibs/TEMPLATE.md`** had a dangling reference to a non-existent `TIB-2026-04-08`.
8. **`web3-security` persona** was the last one still in the old "Prompt must include:" bullet-list format — never got the structural refactor the others received.
9. **Release-commit signing patterns from PR #673** (release-bot hardening: GitHub-signed identity via `createCommitOnBranch`, write-token guards) weren't in §10 — would land without spec backing.

### A. Expanded `documentation` persona

Old scope: JSDoc on exported symbols per `docs/jsdoc-style.md`.

New scope adds:

- **Markdown documentation accuracy** across `README.md`, `AGENTS.md`, `MISSION.md`, `CONTRIBUTING.md`, `SECURITY.md`, `docs/**`, `.agents/**`, package `*.md` files. Flags stale prose, out-of-sync inventories, cross-doc consistency drift, code blocks that no longer match the code they describe.
- **Pointer / link integrity** — internal Markdown links, path references in prose, frontmatter references (`applies:`, `trigger:`, `canonical-rules:`). Rename cascade detection via `git diff --name-status --find-renames`.
- **AGENTS.md ↔ persona backlink consistency** — every persona's `applies:` must match the corresponding `> Applied by personas: …` callout in `AGENTS.md`. Dangling anchors and one-way pointers are findings.

### B. New `AGENTS.md` §10 — Review automation & CI/release security

- Canonical persona inventory (baseline + conditional tables) with `applies:` columns linking each persona back to spec.
- **Orchestration** sub-table for the three command skills (`.agents/commands/pr-review-{ci,gh,local}.md`) + the shared base (`.agents/lib/pr-review-base.md`) — the full review-automation stack is now discoverable.
- Source-of-truth CI/release security rules: workflow injection, `pull_request_target` + checkout restriction, ACL-gated comment triggers, action pinning, `permissions:` block, secret exposure, publish-flow integrity, Changesets/release-bot wiring, lockfile drift, dependency hygiene, `.npmrc` hardening, **and (commit `de2075ce`) release-commit signing & write-token hardening** (informed by PR #673).
- `ci-release-security` references §10 instead of being authoritative on its own.

### C. Backlinks in `AGENTS.md` §3–8

Each existing section ends with a `> Applied by personas: …` callout. Pairs with the new `applies:` frontmatter for bidirectional navigation.

### D. Persona frontmatter `applies:` + `out-of-scope:`

All 8 personas declare `applies:` (spec anchor) and `out-of-scope:` (boundary against neighbors).

### E. Persona dedupe (audit-driven)

- **`module-api-architecture`** — 8 verbatim restatements of AGENTS.md §1/§3/§4/§8 replaced with section pointers + crisp application points.
- **`style-conventions`** — verbatim §3/§8 restatements replaced with pointers. Scope sharpened to "mechanical-style enforcement + changeset relevance".
- **`code-quality`** + **`silent-failure-hunter`** boundary sharpened: `code-quality` owns *existence* of typed error classes (§2); `silent-failure-hunter` owns *depth* of error handling. `code-quality.applies:` shifted from §1/§3 to §2/§3.
- **`web3-security`** (last persona in the old format) — rewritten with structured What-to-flag sections (contract interaction shape, transaction parameter integrity, wallet/chain handling, permit/typed-data, token approvals, race conditions, Action-layer purity), severity guidance calibrated for the domain, sharpened `out-of-scope` boundary against `code-quality` (which owns generic security primitives).

### F. `test-coverage` rewrite — colocation enforcer

Reflects AGENTS.md §5: new unit tests live next to source (`src/Foo.ts` ↔ `src/Foo.test.ts`). Two layouts coexist (colocation wired in `morpho-sdk` + `evm-simulation` today; older `packages/<pkg>/test/` layout elsewhere). Fork-bound tests use `*.integration.test.ts` naming. Refactor-driven migration path documented. Forward-reference to `@morpho-org/test/mock` marked as "lands with TIB-2026-04-27" (PR #596).

### G. CONTRIBUTING.md trim

"Code Style" + "Changesets" sections collapsed to brief operational quick-reference + pointers to `AGENTS.md` §7/§8.

### H. Housekeeping

- `docs/tibs/TEMPLATE.md` footer: removed dangling `TIB-2026-04-08` reference; added explicit notes that TIBs are *proposals feeding into AGENTS.md* (not authoritative themselves) and that the repo uses two ID formats (sequential for older TIBs, CalVer for new).

## Audit verdict on the rest of the repo

Spawned an Explore agent to survey docs not covered by the rounds above:
- Per-package `AGENTS.md` (16 packages), nested `AGENTS.md` under `morpho-sdk/src/`, package READMEs, `ARCHITECTURE.md`, `BUNDLER3.md`, remaining `.agents/commands/*` (`pr-create`, `pr-describe`, `pr-fix`, `extract-plan`, `verify-dependency-vulnerability`).
- **All clean** — no drift from the refactor, no stale references, no broken backlinks. Inheritance chain (root → package → nested) intact across the SDK tree.

## Scope discipline

- **No published package source touched** → no changeset per `CLAUDE.md` §7.
- Skill / command files untouched (delegation pattern unchanged).
- TIB-0002 and TIB-0003 left alone — they are `Status: Proposed` (live decisions, not stale).
- Deferred deliberately (separate PRs): JSDoc backfill (tracked in TIB-2026-05-04), legacy `test/` → colocation migration for non-wired packages, `extract-plan.md` Team-IDs table (gated on Linear integration).

## Test plan

- [ ] Every `> Applied by personas: …` callout in `AGENTS.md` links to a real file (verified — 8/8 resolve).
- [ ] Every persona's `applies:` line cites a real `AGENTS.md` section (verified — 8/8 frontmatter + 7 backlinks consistent).
- [ ] Persona relative paths (`../../AGENTS.md`, `../../MISSION.md`) resolve from `.agents/personas/` (verified).
- [ ] `documentation` persona body references files that exist (`docs/jsdoc-style.md`, `.agents/lib/pr-review-base.md` — verified). Forward-references to `@morpho-org/test/mock` correctly marked as incoming.
- [ ] Run `/pr-review-local` on this branch — eat our own dogfood; the expanded `documentation` persona should find no pointer drift, and the new `web3-security` / `ci-release-security` rule structures should produce findings cleanly. (Done — see follow-up commit if fixes applied.)

## Commit history

| SHA | Title |
|---|---|
| `8a6b8ff2` | Initial four changes — documentation persona expansion, §10, backlinks, frontmatter |
| `300cfbe6` | Audit dedupe — 5 persona refactors + CONTRIBUTING.md trim + TEMPLATE.md footer |
| `21f6ae12` | Self-test fix — forward-reference to `@morpho-org/test/mock` softened |
| `de2075ce` | Web3-security sharpening + release-signing rule (informed by PR #673) |
| (pending) | Self-applied review fixes (from running `/pr-review-local --fix` on this branch) |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[View Slack thread](https://morpho-labs.slack.com/archives/C0B1C0ULKMK/p1779183051494739)

### CODE REVIEW

File: .agents/lib/pr-review-base.md [hunk 1]
```diff
@@ -78,7 +78,15 @@ For each unique package directory among the changed files (e.g. a file at `packa
 
 Use the Glob tool: `**/AGENTS.md` and `packages/*/*.md`. Filter to paths that prefix at least one changed file's directory. Files outside `packages/` use only items 1–4 of the root baseline (items 5–7 conditional as triggered).
 
-After discovery, **print** the list of files read so the user can spot omissions:
+### Detect conditional persona triggers
+
+Compute flags from the changed-files list. These flags drive which `kind: conditional` personas launch in Step 5:
+
+- `<HAS_CI_RELEASE>` — true if any changed file matches `.github/workflows/**`, `.github/actions/**`, `.changeset/**`, root or package `package.json` (when a `scripts.*publish*` / `scripts.*release*` field is touched), `pnpm-lock.yaml`, `pnpm-workspace.yaml`, `.npmrc`, OR if any changed file contains `changeset publish`, `npm publish`, `pnpm publish`, or `gh release create`.
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply .agents/lib/pr-review-base.md:85]: > Trigger CI persona for dependency edits — `<HAS_CI_RELEASE>` stays false when a PR edits `package.json` deps without a matching lockfile change, so the persona never runs for dependency-hygiene rules in that case.

Skipping in this PR — the trigger expansion is a legitimate suggestion but it's a substantive behavior change (the persona would fire on every dep-bump PR even when the lockfile is correctly updated, flooding with no-op runs) and goes beyond this PR's doc-architecture scope.

Better as a follow-up that reviews the trigger surface holistically — likely with an additional, narrower flag like `<HAS_DEPENDENCY_CHANGE>` so dependency-hygiene rules can fire without flooding on every routine dep bump. Leaving open for human review.

File: .agents/lib/pr-review-base.md [hunk 2]
```diff
@@ -266,5 +191,6 @@ The caller (Step 7 of `/pr-review-ci` / `/pr-review-gh` / `/pr-review-local`) co
 - `<FINDINGS>` — sorted, deduplicated array of `{severity, file, line, description}`.
 - `<FAILED_AGENTS>` — count + names of agents that returned `agent_error` or malformed output.
 - `<COUNTS>` — `{critical, high, medium, low}` totals.
+- `<TOTAL_AGENTS_LAUNCHED>` — count of personas that actually fired (baseline always-fire count + any conditional personas whose trigger flag was true). Used by the caller's report to phrase "<FAILED_AGENTS> of <TOTAL_AGENTS_LAUNCHED> agents failed" correctly when conditional personas did not fire.
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply .agents/lib/pr-review-base.md:194]: > Wire callers to the dynamic agent count — `<TOTAL_AGENTS_LAUNCHED>` is dynamic but caller files still hardcode `of 7 agents`.

Fixed in 4ea0d581 — replaced all 9 occurrences of `of 7 agents` with `of <TOTAL_AGENTS_LAUNCHED> agents` across `pr-review-{ci,gh,local}.md`.

File: .agents/personas/test-coverage.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+Fork-bound tests use **`*.integration.test.ts`** naming when they stay (e.g. a test that depends on a real Anvil fork's contract state). This lets unit-only test runs filter them out cleanly.
+
+## What to flag
+
+### Missing coverage (the always-applies set)
+
+- New public exports without a corresponding test — in colocated layout (`src/Foo.test.ts` next to `src/Foo.ts`) for packages wired for colocation, or `packages/<pkg>/test/Foo.test.ts` for the older layout.
+- New code paths inside existing exports without test cases — branches, error paths, edge cases like `0n`, `MAX_UINT256`, negative `bigint`, empty arrays, NaN-equivalents.
+- Removed or modified public exports without their tests updated (e.g. signature change, behavior change).
+- Onchain code paths (any code calling `viem` / `wagmi` actions) — confirm at least one test exercises the path. Mocked transport-level is preferred for new unit tests (the `@morpho-org/test/mock` sub-export with `createMockClient` / `mockRead` / `expectReadCall` lands with TIB-2026-04-27); fork-based is acceptable for path-purpose tests, and those should use `*.integration.test.ts` naming.
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply .agents/personas/test-coverage.md]: > Avoid recommending unavailable mock tests — the persona names `@morpho-org/test/mock` and recommends mocked transports as preferred, but AGENTS.md §5 still says fork-based, and the sub-export doesn't exist yet (lands with TIB-2026-04-27).

Fixed in 4ea0d581 — reworded the test-coverage persona to defer to AGENTS.md §5 (forks via `@morpho-org/test`) today, and clearly mark the mocked-transport convention as gated on TIB-2026-04-27 (PR #596) acceptance. Until that TIB lands and §5 is amended, the persona will NOT recommend mocked transports.

File: AGENTS.md [hunk 4]
```diff
@@ -168,3 +182,57 @@ A scannable list of patterns reviewers reject. Most are review-only today (per t
...
+| [`.agents/commands/pr-review-{ci,gh,local}.md`](./.agents/commands/) | Three caller-side skills — one per use case (CI verdict / local PR / pre-PR terminal). Each parses args, resolves branches, then delegates Steps 3–6 to the shared base. |
+| [`.agents/lib/pr-review-base.md`](./.agents/lib/pr-review-base.md) | Shared review base — reads `<PROJECT_CONTEXT>`, computes conditional flags, loops over personas, aggregates and deduplicates findings. Does not encode any rule itself; it only orchestrates. |
+
+### Persona inventory
+
+Baseline personas (always fire):
+
+| Persona | Anchors | Focus |
+|---|---|---|
+| [`code-quality`](./.agents/personas/code-quality.md) | §1, §3 | Type safety, code smells, naming, cross-file impact on SDK consumers, security primitives. |
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply AGENTS.md]: > §10 inventory table lists wrong AGENTS.md anchors for code-quality persona (§1,§3 instead of §2,§3).

Already addressed in commit d734a65e (the self-applied `/pr-review-local --fix` pass also caught this). The §10 row now reads `§2, §3` matching the persona's `applies:`.

File: .agents/personas/module-api-architecture.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
+---
+name: module-api-architecture
+kind: baseline
+applies: AGENTS.md §1 Architecture (layering, modularity), §4 Public API & packaging
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply .agents/personas/module-api-architecture.md]: > §3 backlink lists module-api-architecture, but persona's `applies:` doesn't claim §3 — bidirectional drift.

Fixed in 4ea0d581 — added `§3 Type discipline (at the boundary)` to `module-api-architecture.md`'s `applies:` frontmatter, and updated the AGENTS.md §10 inventory row to list `§1, §3, §4`. Persona body already enforced §3 type-discipline rules at the boundary; now the frontmatter agrees.

File: AGENTS.md [hunk 6]
```diff
@@ -95,6 +99,8 @@ A scannable list of patterns reviewers reject. Most are review-only today (per t
 - Each package has one reason to exist (see §1 Modularity). Framework adapters never live in core SDK packages.
 - Don't replace a small local helper with a transitive dep just to "reuse code".
 
+> Applied by personas: [`module-api-architecture`](./.agents/personas/module-api-architecture.md), [`style-conventions`](./.agents/personas/style-conventions.md).
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply AGENTS.md]: > §4 backlink lists style-conventions, but persona's `applies:` doesn't claim §4 — and the body covers Biome mechanics + changesets, not §4's public-API packaging rules.

Fixed in 4ea0d581 — dropped style-conventions from AGENTS.md §4's `> Applied by` callout. §4 now correctly lists only module-api-architecture (which now declares §3 + §4 in its frontmatter — see the sibling thread).

File: AGENTS.md [hunk 7]
```diff
@@ -168,3 +182,57 @@ A scannable list of patterns reviewers reject. Most are review-only today (per t
...
+### Persona inventory
+
+Baseline personas (always fire):
+
+| Persona | Anchors | Focus |
+|---|---|---|
+| [`code-quality`](./.agents/personas/code-quality.md) | §1, §3 | Type safety, code smells, naming, cross-file impact on SDK consumers, security primitives. |
+| [`module-api-architecture`](./.agents/personas/module-api-architecture.md) | §1, §4 | Package boundaries, public surface, NodeNext import discipline. |
+| [`web3-security`](./.agents/personas/web3-security.md) | §1 (Action layer), §2 | Contract interactions, transaction params, permit flows, race conditions. Severity defaults to critical/high. |
+| [`silent-failure-hunter`](./.agents/personas/silent-failure-hunter.md) | §1 (testability), §2 | Swallowed errors, missing error states, dead code paths. |
```
[the remark this answers is not in the document]
Comment (0xbulma) [reply AGENTS.md]: > §1 and §2 have no `> Applied by personas:` callouts despite multiple personas claiming them.

Already addressed in commit d734a65e (the self-applied `/pr-review-local --fix` pass). §1 now lists module-api-architecture, web3-security (Action-layer purity), silent-failure-hunter (testability). §2 lists code-quality (forbidden patterns 1–4, 7–10), web3-security (signing in builders, rule 3), silent-failure-hunter (typed errors, rule 2).

File: .agents/lib/pr-review-base.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xbulma) [reply .agents/lib/pr-review-base.md:194]: > Caller files hardcode 'of 7 agents' but base outputs dynamic `<TOTAL_AGENTS_LAUNCHED>`.

Fixed in 4ea0d581 — same as the sibling Codex thread. Replaced all 9 occurrences across `pr-review-{ci,gh,local}.md` with the dynamic placeholder.

**Review by Rubilmax — APPROVED** [review]
Exciting!

**Comment by Foulks-Plb** [comment]
Internal contradictions:

§2 backlink contradicts the personas it cites. §2 credits silent-failure-hunter for “2 — typed errors”, but that persona’s own body + out-of-scope: explicitly defer the typed-class existence rule to code-quality. Also missing: module-api-architecture for #5 (deep cross-package imports) and test-coverage/web3-security for #6 (mocked viem clients) — “1–4, 7–10” leaves both orphaned.

§1 backlink ↔ code-quality.md disagree on early returns. code-quality.md labels “Prefer early returns over nested conditionals” as “not codified in AGENTS.md”, but it’s the last bullet of §1 “Stateless, immutable, composable”. Either add code-quality to §1’s backlink, or drop the “not codified” claim.

The dedupe the PR claims wasn’t done for CI/release. Motivation §4 (“personas restate AGENTS.md rules verbatim — sync burden + drift”) is violated twice:

<HAS_CI_RELEASE> trigger is restated verbatim in 3 places (pr-review-base.md Step 4, ci-release-security.md Trigger, AGENTS.md §10 table). Pick one canonical (probably Step 4 since it computes the flag) and point at it.
CI/release rules are duplicated between §10 and ci-release-security.md, and they already diverge: persona has “unknown publishers” and auto-install-peers rules that §10 doesn’t; wording on secrets: inherit differs. If §10 is source-of-truth as the PR states, the persona should reference it instead of paraphrasing.
silent-failure-hunter.applies: cites §1 (testability) with no real anchor. §1 “Testability” is about DI / I/O at the edges — nothing about error swallowing. The persona doesn’t enforce anything from that subsection. Drop §1 from applies: or narrow it to §2.
